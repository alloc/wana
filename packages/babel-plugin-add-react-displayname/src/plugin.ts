import type { NodePath, PluginPass, types as t } from '@babel/core'
import { declare } from '@babel/helper-plugin-utils'

type Options = {
  callees?: string[]
}

export default declare(babel => {
  babel.assertVersion(7)

  const { types: t } = babel

  const FunctionVisitor: babel.Visitor<PluginPass> = {
    FunctionExpression(path, state) {
      setDisplayNameIfPossible(path, state)
    },
    ArrowFunctionExpression(path, state) {
      setDisplayNameIfPossible(path, state)
    },
  }

  return {
    name: '@wana/babel-plugin-add-react-displayname',
    visitor: {
      VariableDeclaration(path, state) {
        path.traverse(FunctionVisitor, state)
      },
    },
  }

  function setDisplayNameIfPossible(
    path: NodePath<t.FunctionExpression | t.ArrowFunctionExpression>,
    state: PluginPass
  ) {
    const prevStmt = findBlockChild(path)
    if (!prevStmt) return

    const displayName = inferDisplayName(path, state.opts as Options)
    if (!displayName || /^[$_]?[a-z]/.test(displayName)) return

    if (path.isArrowFunctionExpression()) {
      const { params, body, returnType, typeParameters } = path.node
      const bodyBlock = t.isExpression(body)
        ? t.blockStatement([t.returnStatement(body)])
        : body
      const functionExpr = t.functionExpression(
        t.identifier(displayName),
        params,
        bodyBlock,
        false,
        false
      )
      functionExpr.returnType = returnType
      functionExpr.typeParameters = typeParameters
      path.replaceWith(functionExpr)
    } else {
      const newStmt = t.expressionStatement(
        t.assignmentExpression(
          '=',
          t.memberExpression(
            t.identifier(displayName),
            t.identifier('displayName')
          ),
          t.stringLiteral(displayName)
        )
      )
      newStmt.trailingComments = prevStmt.node.trailingComments
      prevStmt.node.trailingComments = null
      prevStmt.insertAfter(newStmt)
    }
  }

  function findLeftValue(expr: NodePath<t.Expression>) {
    let path = expr.parentPath
    while (path) {
      if (path.isBlockParent() || path.isProperty()) {
        return
      }
      if (path.isAssignmentExpression()) {
        return path.node.left
      }
      if (path.isVariableDeclarator()) {
        return path.node.id
      }
      path = path.parentPath!
    }
  }

  function findBlockChild(path: NodePath) {
    while (path.parentPath) {
      if (path.parentPath.isProperty()) {
        return
      }
      if (path.parentPath.isBlockParent()) {
        return path
      }
      path = path.parentPath
    }
  }

  function inferDisplayName(
    funcPath: NodePath<t.FunctionExpression | t.ArrowFunctionExpression>,
    opts: Options
  ) {
    const parentNode = funcPath.parentPath.node
    if (t.isVariableDeclarator(parentNode)) {
      // Ignore plain function components, like "const Foo = () => { ... }"
      return null
    }
    if (t.isCallExpression(parentNode) && parentNode.callee == funcPath.node) {
      // Ignore IIFEs that return JSX, like "(() => { ... })()"
      return null
    }
    if (
      doesReturnJSX(funcPath.node.body) ||
      (opts.callees && opts.callees.some(callee => hasCallee(funcPath, callee)))
    ) {
      const left = findLeftValue(funcPath)
      if (t.isIdentifier(left)) {
        return left.name
      }
    }
    return null
  }

  function hasCallee(path: NodePath, callee: string) {
    path = path.parentPath!
    while (path.isCallExpression()) {
      if (t.isIdentifier(path.node.callee, { name: callee })) {
        return true
      }
      path = path.parentPath
    }
    return false
  }

  function doesReturnJSX(node: t.Expression | t.BlockStatement) {
    if (!node) return false
    if (t.isJSX(node)) {
      return true
    }

    if (t.isBlockStatement(node)) {
      const len = node.body.length
      if (len) {
        const lastNode = node.body[len - 1]
        if (t.isReturnStatement(lastNode)) {
          return lastNode.argument && t.isJSX(lastNode.argument)
        }
      }
    }

    return false
  }
})
