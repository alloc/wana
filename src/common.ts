export const emptyArray: readonly any[] = Object.freeze([])

export const setHidden = (obj: any, key: any, value: any) =>
  Object.defineProperty(obj, key, { value, writable: true, configurable: true })

export const isObject = (value: unknown): value is object =>
  value && typeof value == 'object'

export const isFunction = (value: unknown): value is Function =>
  typeof value == 'function'

export function rethrowError(error: Error) {
  throw error
}

export const hasOwn = Function.call.bind({}.hasOwnProperty) as (
  obj: object,
  key: keyof any
) => boolean

export const noop = () => {}

export const nope = () => false

export const todo: any = () => {
  throw Error('Not yet implemented')
}
