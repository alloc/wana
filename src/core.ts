// Observation
export { o } from './o'
export { auto, Auto } from './auto'
export { when } from './when'
export { watch, Watcher } from './watch'
export { Observer, Change } from './observable'

// Escape hatches
export { no } from './no'
export { noto } from './noto'

// Internals
export { $O, $$ } from './symbols'
