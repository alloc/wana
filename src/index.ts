export { o, Observer, Change } from './observable'
export { auto, Auto } from './auto'
export { withAuto, useAuto, useO } from './render'
