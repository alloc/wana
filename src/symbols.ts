/** For accessing the proxied state */
export const $$ = Symbol.for('wana:state')

/** For storing an `Observable` instance, and observing all events */
export const $O = Symbol.for('wana:observable')

/** For storing an observable proxy */
export const $P = Symbol.for('wana:proxy')

/** For observing "size" changes */
export const SIZE = Symbol.for('wana:size')
