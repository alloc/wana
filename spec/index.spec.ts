import { $$, $O, auto, flushSync, globals, no, o } from 'wana/core'
import { SIZE } from '../src/symbols'

const observed: any[] = []
afterEach(() => {
  observed.length = 0
})

const track = (effect: () => void) => {
  globals.observe = (target, key) => observed.push([target, key])
  effect()
  globals.observe = null
}

const watch = (obj: object) => {
  const spy = jest.fn()
  const observer = { onChange: spy }
  obj[$O].get($O).add(observer)
  obj[$O].get(SIZE).add(observer)
  return spy.mock.calls
}

describe('o()', () => {
  it('never creates more than one proxy per object', () => {
    const obj = {}
    expect(o(obj)).toBe(o(obj))
    expect(o(obj)).not.toBe(obj)
  })

  it('is a no-op for primitives', () => {
    expect(o(0)).toBe(0)
    expect(o(true)).toBe(true)
    expect(o(null)).toBe(null)
    expect(o(undefined)).toBe(undefined)
  })

  it('is a no-op for certain built-in types', () => {
    const date = new Date()
    expect(o(date)).toBe(date)

    const regExp = /./g
    expect(o(regExp)).toBe(regExp)

    const nativePromise = Promise.resolve()
    expect(o(nativePromise)).toBe(nativePromise)

    const generatorFn = function*() {}
    expect(o(generatorFn)).toBe(generatorFn)

    const generator = generatorFn()
    expect(o(generator)).toBe(generator)

    const asyncFn = async () => {}
    expect(o(asyncFn)).toBe(asyncFn)

    const weakMap = new WeakMap()
    expect(o(weakMap)).toBe(weakMap)

    const weakSet = new WeakSet()
    expect(o(weakSet)).toBe(weakSet)
  })

  it('is a no-op for frozen objects', () => {
    const obj = Object.freeze({})
    expect(o(obj)).toBe(obj)
  })

  describe('property getter', () => {
    it('is called with the proxy as its "this" value', () => {
      class Foo {
        get foo() {
          return this
        }
      }
      const foo = o(new Foo())
      expect(foo.foo).toBe(foo)
    })
  })

  describe('property setter', () => {
    it('is called with the proxy as its "this" value', () => {
      class Foo {
        this: any
        get foo() {
          return this
        }
        set foo(out: Foo) {
          out.this = this
        }
      }
      const foo = o(new Foo())
      foo.foo = foo
      expect(foo.this).toBe(foo)
    })
  })
})

describe('o(Object)', () => {
  describe('[[Get]]', () => {
    it('works with no observer', () => {
      const obj: any = o({ a: 1 })
      expect(obj.a).toBe(1)
      expect(obj.b).toBeUndefined()
    })

    it('tracks known keys', () => {
      const obj: any = o({ a: 1 })
      track(() => obj.a)
      expect(observed.length).toBe(1)
      expect(observed[0][0]).toBe(obj[$$])
      expect(observed[0][1]).toBe('a')

      obj.b = 1
      track(() => obj.b)
      expect(observed.length).toBe(2)
      expect(observed[1][0]).toBe(obj[$$])
      expect(observed[1][1]).toBe('b')
    })

    it('tracks unknown keys', () => {
      const obj: any = o({})
      track(() => obj.a)
      expect(observed.length).toBe(1)
      expect(observed[0][0]).toBe(obj[$$])
      expect(observed[0][1]).toBe('a')
    })

    it('ignores inherited keys', () => {
      const obj: any = o({})
      track(() => obj.toString)
      expect(obj.toString).toBeDefined()
      expect(observed).toEqual([])
    })

    describe('own getter', () => {
      it('is called with observable this', () => {
        const obj = o({
          get a() {
            expect(this).toBe(obj)
            return 0
          },
        })
        obj.a // tslint:disable-line
        expect.assertions(1)
      })
    })

    describe('inherited getter', () => {
      it('is called with observable this', () => {
        const obj: any = o({
          __proto__: {
            get a() {
              expect(this).toBe(obj)
              return 0
            },
          },
        })
        obj.a // tslint:disable-line
        expect.assertions(1)
      })
    })
  })

  describe('[[Set]]', () => {
    it('emits for new keys', () => {
      const obj: any = o({})
      const calls = watch(obj)
      obj.a = 1
      expect(calls[0][0].target).toBe(obj[$$])
      expect(calls).toMatchSnapshot()
    })

    it('emits for new values', () => {
      const obj: any = o({ a: 0 })
      const calls = watch(obj)
      obj.a = 1
      expect(calls[0][0].target).toBe(obj[$$])
      expect(calls).toMatchSnapshot()
    })

    it('skips emit for old values', () => {
      const obj: any = o({ a: 1 })
      const calls = watch(obj)
      obj.a = 1
      expect(calls).toEqual([])
    })

    it('throws when a non-writable property is assigned to', () => {
      const obj: any = o({})
      Object.defineProperty(obj, 'a', { value: 1 })
      expect(() => {
        obj.a = 2
      }).toThrowErrorMatchingInlineSnapshot(
        `"Cannot assign to read only property 'a' of object '#<Object>'"`
      )
    })

    it('throws when a getter is assigned to', () => {
      const obj: any = o({})
      Object.defineProperty(obj, 'a', { get: () => 1 })
      expect(() => {
        obj.a = 2
      }).toThrowErrorMatchingInlineSnapshot(
        `"'set' on proxy: trap returned falsish for property 'a'"`
      )
    })

    describe('own setter', () => {
      it('is called with observable this', () => {
        const obj = o({
          get a() {
            return 0
          },
          set a(_val) {
            expect(this).toBe(obj)
          },
        })
        obj.a = 1
        expect.assertions(1)
      })

      it('can be observed', () => {
        const obj = o({
          get a() {
            return 0
          },
          set a(_val) {
            expect(globals.observe).toBeTruthy()
          },
        })
        auto(() => {
          obj.a = 1
        })
        expect.assertions(1)
      })
    })

    describe('inherited setter', () => {
      it('is called with observable this', () => {
        const obj: any = o({
          __proto__: {
            get a() {
              return 0
            },
            set a(_val) {
              expect(this).toBe(obj)
            },
          },
        })
        obj.a = 1
        expect.assertions(1)
      })

      it('can be observed', () => {
        const obj: any = o({
          __proto__: {
            get a() {
              return 0
            },
            set a(_val) {
              expect(globals.observe).toBeTruthy()
            },
          },
        })
        auto(() => {
          obj.a = 1
        })
        expect.assertions(1)
      })
    })
  })

  describe('[[Delete]]', () => {
    it('emits for known keys', () => {
      const obj: any = o({ a: 1 })
      const calls = watch(obj)
      delete obj.a
      expect(calls[0][0].target).toBe(obj[$$])
      expect(calls).toMatchSnapshot()
    })

    it('skips emit for unknown keys', () => {
      const obj: any = o({})
      const calls = watch(obj)
      delete obj.a
      expect(calls).toEqual([])
    })
  })

  describe('"in" operator', () => {
    it.todo('tracks known keys')
    it.todo('tracks unknown keys')
    it.todo('tracks inherited keys')
  })

  describe('[[DefineOwnProperty]]', () => {
    it('emits for new property', () => {
      const obj: any = o({})
      const calls = watch(obj)

      // 1. Define with "get" descriptor
      Object.defineProperty(obj, 'a', {
        get: () => 1,
      })

      // 2. Define with "value" descriptor
      Object.defineProperty(obj, 'b', {
        value: 2,
      })

      expect(calls).toMatchSnapshot()
    })

    it('emits for redefined property', () => {
      const obj = o({
        get a() {
          return 1
        },
      })
      const calls = watch(obj)

      // 1. Redefine with "get" descriptor
      Object.defineProperty(obj, 'a', {
        get: () => 2,
      })

      // 2. Redefine with "value" descriptor
      Object.defineProperty(obj, 'a', {
        value: 3,
        writable: true,
      })

      // 3. Redefine with "value" descriptor (again)
      Object.defineProperty(obj, 'a', {
        value: 4,
      })

      expect(calls).toMatchSnapshot()
    })
  })
})

describe('o(Array)', () => {
  it('works with Array.isArray', () => {
    const arr = o([])
    expect(Array.isArray(arr)).toBeTruthy()
  })

  describe('[[Get]]', () => {
    it('ignores known indices', () => {
      const arr = o([0])
      track(() => arr[0])
      expect(observed).toEqual([])

      arr[2] = 2
      track(() => arr[2])
      expect(observed).toEqual([])
    })

    it('ignores unknown indices', () => {
      const arr = o([])
      track(() => arr[0])
      expect(observed).toEqual([])
    })
  })

  describe('[[Set]]', () => {
    it('emits for new indices', () => {
      const arr: any[] = o([])
      const calls = watch(arr)
      arr[1] = 1
      expect(calls[0][0].target).toBe(arr[$$])
      expect(calls).toMatchSnapshot()
    })

    it('emits for new values', () => {
      const arr: any[] = o([0])
      const calls = watch(arr)
      arr[0] = 1
      expect(calls[0][0].target).toBe(arr[$$])
      expect(calls).toMatchSnapshot()
    })

    it('skips emit for old values', () => {
      const arr: any[] = o([1])
      const calls = watch(arr)
      arr[0] = 1
      expect(calls).toEqual([])
    })
  })

  describe('.length', () => {
    it('can be tracked', () => {
      const arr = o([])
      track(() => arr.length)
      expect(observed.length).toBe(1)
      expect(observed[0][0]).toBe(arr[$$])
      expect(observed[0][1]).toBe(SIZE)
    })

    it('emits when set directly', () => {
      const arr = o([])
      const calls = watch(arr)
      // No "splice" event is emitted when increasing length like this,
      // because all of the indices are holes:
      arr.length = 3
      expect(calls.length).toBe(1)
      // Emit a "splice" event here, even if all of the truncated indices
      // are holes, since checking for holes isn't worth the extra logic:
      arr.length = 0
      expect(calls[0][0].target).toBe(arr[$$])
      expect(calls).toMatchSnapshot()
    })

    it('skips emit when set to same length', () => {
      const arr = o([1])
      const calls = watch(arr)
      arr.length = 1
      expect(calls).toEqual([])
    })
  })

  describe('.splice()', () => {
    describe('removal', () => {
      it('can splice the start', () => {
        const arr = o([1, 2, 3])
        const calls = watch(arr)
        arr.splice(0, 2)
        expect(calls).toMatchSnapshot()
      })

      it('can splice the middle', () => {
        const arr = o([1, 2, 3, 4])
        const calls = watch(arr)
        arr.splice(1, 2)
        expect(calls).toMatchSnapshot()
      })

      it('can splice the end', () => {
        const arr = o([1, 2, 3])
        const calls = watch(arr)
        arr.splice(1, 2)
        expect(calls).toMatchSnapshot()
      })
    })

    describe('insertion', () => {
      it('can insert at the start', () => {
        const arr = o([1])
        const calls = watch(arr)
        arr.splice(0, 0, 3, 2)
        expect(calls).toMatchSnapshot()
      })

      it('can insert into the middle', () => {
        const arr = o([1, 2])
        const calls = watch(arr)
        arr.splice(1, 0, 0, 0)
        expect(calls).toMatchSnapshot()
      })

      it('can insert at the end', () => {
        const arr = o([1])
        const calls = watch(arr)
        arr.splice(1, 0, 2, 3)
        expect(calls).toMatchSnapshot()
      })
    })

    describe('remove count', () => {
      it('removes nothing when zero or less', () => {
        const arr = o([1, 2])
        const calls = watch(arr)
        arr.splice(0, 0)
        arr.splice(1, -1)
        expect(calls).toEqual([])
      })

      it('can exceed the last index', () => {
        const arr = o([1, 2])
        const calls = watch(arr)
        arr.splice(1, 3)
        expect(calls).toMatchSnapshot()
      })
    })

    describe('negative start index', () => {
      it('starts from the end', () => {
        const arr = o([1, 2, 3])
        const calls = watch(arr)
        arr.splice(-2, 1)
        expect(calls).toMatchSnapshot()
      })

      it('stops at zero', () => {
        const arr = o([1, 2])
        const calls = watch(arr)
        arr.splice(-100, 1)
        expect(calls).toMatchSnapshot()
      })
    })

    it('can insert and remove at the same time', () => {
      const arr = o([1, 2, 3])
      const calls = watch(arr)
      // more insertions
      arr.splice(1, 1, 0, 0)
      expect(calls).toMatchSnapshot()
      calls.length = 0
      // more removals
      arr.splice(1, 2, 3)
      expect(calls).toMatchSnapshot()
      calls.length = 0
      // equal amounts
      arr.splice(1, 1, 0)
      expect(calls).toMatchSnapshot()
    })

    // To avoid extra logic for rare edge cases:
    it('emits even if values did not change', () => {
      const arr = o([1, 2, 3])
      const calls = watch(arr)
      arr.splice(0, arr.length, ...arr)
      expect(calls).toMatchSnapshot()
    })
  })

  describe('.push()', () => {
    it.todo('emits any changes')
    it.todo('skips emit when nothing is added')
  })

  describe('.unshift()', () => {
    it.todo('emits any changes')
    it.todo('skips emit when nothing is added')
  })

  describe('.shift()', () => {
    it.todo('emits any changes')
    it.todo('skips emit if empty')
  })

  describe('.pop()', () => {
    it.todo('emits any changes')
    it.todo('skips emit if empty')
  })

  describe('.reverse()', () => {
    it.todo('emits any changes')
    // To avoid extra logic for rare edge cases:
    it.todo('emits even if nothing changes')
  })

  describe('.sort()', () => {
    it.todo('emits any changes')
    // To avoid extra logic for rare edge cases:
    it.todo('emits even if nothing changes')
  })
})

describe('o(Set)', () => {
  describe('.has()', () => {
    it.todo('observes the entire set')
  })

  describe('.add()', () => {
    it('emits for new values', () => {
      const set = o(new Set())
      const calls = watch(set)
      set.add(1)
      expect(calls[0][0].target).toBe(set[$$])
      expect(calls).toMatchSnapshot()
    })

    it('skips emit for old values', () => {
      const set = o(new Set([1]))
      const calls = watch(set)
      set.add(1)
      expect(calls).toEqual([])
    })
  })

  describe('.delete()', () => {
    it('emits for known values', () => {
      const set = o(new Set([1]))
      const calls = watch(set)
      set.delete(1)
      expect(calls[0][0].target).toBe(set[$$])
      expect(calls).toMatchSnapshot()
    })

    it('skips emit for unknown values', () => {
      const set = o(new Set())
      const calls = watch(set)
      set.delete(1)
      expect(calls).toEqual([])
    })
  })

  describe('.clear()', () => {
    it('emits only when not empty', () => {
      const set = o(new Set())
      const calls = watch(set)
      set.clear()
      expect(calls).toEqual([])
      set.add(1).add(2)
      calls.length = 0
      set.clear()
      expect(calls).toMatchSnapshot()
    })
  })
})

describe('o(Map)', () => {
  describe('.has()', () => {
    it.todo('works with no observer')
    it.todo('tracks known keys')
    it.todo('tracks unknown keys')
  })

  describe('.get()', () => {
    it.todo('works with no observer')
    it.todo('tracks known keys')
    it.todo('tracks unknown keys')
  })

  describe('.set()', () => {
    it.todo('emits for known keys')
    it.todo('emits for unknown keys')
    it.todo('skips emit if unchanged')
  })

  describe('.delete()', () => {
    it.todo('emits for known keys')
    it.todo('skips emit for unknown keys')
  })

  describe('.clear()', () => {
    it.todo('emits only when not empty')
  })
})

// Note: See "useDerived.spec.tsx" for more tests.
describe('o(Function)', () => {
  it.todo('returns an observable getter')
  it.todo('avoids wrapping an already observable getter')

  it('can be observed by an "auto" call', () => {
    const state = o({ count: 1 })

    const get = jest.fn(() => Math.abs(state.count) * 2)
    const memo = o(get)

    const effect = jest.fn(() => memo())
    const reaction = auto(effect)
    const onDirty = (reaction.onDirty = jest.fn(
      reaction.onDirty.bind(reaction)
    ))

    const expectedCalls = new Map<any, number>()
    const expectCalls = (fn: jest.Mock, count: number) => {
      let sum = expectedCalls.get(fn)
      if (sum == null) sum = 0
      expect(fn).toBeCalledTimes((sum += count))
      expectedCalls.set(fn, sum)
    }

    // The effect is run immediately.
    expectCalls(effect, 1)

    // Initialize the `memo` value.
    expect(memo()).toBe(2)
    expectCalls(get, 1)

    // Flip the count, so its value is changed,
    // but the `memo` value stays the same.
    state.count = -1

    // No reaction should happen yet.
    expectCalls(effect, 0)
    // ...but it should have been batched.
    expectCalls(onDirty, 1)

    // Synchronous derivation should still work.
    expect(memo()).toBe(2)
    expectCalls(get, 1)

    // Flush the batch and no reaction should occur,
    // because the `memo` value is unchanged.
    flushSync()
    expectCalls(effect, 0)

    // Ensure the `memo` value will be different.
    state.count = 2

    // The reaction should not be synchronous.
    expectCalls(effect, 0)
    // ...but it should have been batched.
    expectCalls(onDirty, 1)

    // Synchronous derivation should still work.
    expect(memo()).toBe(4)
    expectCalls(get, 1)

    // The reaction should happen when flushing the batch,
    // because the `memo` value has changed.
    flushSync()
    expectCalls(effect, 1)
  })

  it('can be observed by a sync "auto" call', () => {
    const state = o({ count: 1 })

    const get = jest.fn(() => Math.abs(state.count) * 2)
    const memo = o(get)

    const effect = jest.fn()
    auto(() => effect(memo()), { sync: true })

    expect(effect).toBeCalledWith(2)
    effect.mockReset()

    state.count++
    expect(effect).toBeCalledWith(4)
    effect.mockReset()

    // This calls into `get`, but `memo` stays the same.
    state.count = -state.count
    expect(effect).not.toBeCalled()
  })

  describe('observing another observable getter', () => {
    it('can rerun synchronously if needed', () => {
      const state = o({ count: 1 })

      const dub = jest.fn(() => state.count * 2)
      const dubMemo = o(dub)

      const quad = jest.fn(() => dubMemo() * 2)
      const quadMemo = o(quad)

      // First run
      expect(quadMemo()).toBe(4)

      // Change a dependency
      state.count += 1

      // The affected getter is not called immediately
      expect(dub).toBeCalledTimes(1)

      // Synchronous rerun
      expect(quadMemo()).toBe(8)
      expect(dub).toBeCalledTimes(2)
    })
  })
})

describe('no()', () => {
  it('returns the original object of an observable proxy', () => {
    let orig: any = {}
    expect(no(o(orig))).toBe(orig)
    orig = []
    expect(no(o(orig))).toBe(orig)
    orig = new Set()
    expect(no(o(orig))).toBe(orig)
    orig = new Map()
    expect(no(o(orig))).toBe(orig)
  })

  it('wraps functions to disable implicit observation for future calls', () => {
    const state = o({ a: 1 })
    const increment = no(() => ++state.a)
    track(() => {
      expect(increment()).toBe(2)
    })
    expect(observed).toEqual([])
  })
})
