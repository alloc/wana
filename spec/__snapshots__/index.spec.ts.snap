// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`o(Array) .length emits when set directly 1`] = `
Array [
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 0,
      "op": "replace",
      "target": Array [],
      "value": 3,
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 3,
      "op": "replace",
      "target": Array [],
      "value": 0,
    },
  ],
  Array [
    Object {
      "key": 0,
      "oldValue": Array [
        undefined,
        undefined,
        undefined,
      ],
      "op": "splice",
      "target": Array [],
      "value": Array [],
    },
  ],
]
`;

exports[`o(Array) .splice() can insert and remove at the same time 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        2,
      ],
      "op": "splice",
      "target": Array [
        1,
        0,
        0,
        3,
      ],
      "value": Array [
        0,
        0,
      ],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 3,
      "op": "replace",
      "target": Array [
        1,
        0,
        0,
        3,
      ],
      "value": 4,
    },
  ],
]
`;

exports[`o(Array) .splice() can insert and remove at the same time 2`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        0,
        0,
      ],
      "op": "splice",
      "target": Array [
        1,
        3,
        3,
      ],
      "value": Array [
        3,
      ],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 4,
      "op": "replace",
      "target": Array [
        1,
        3,
        3,
      ],
      "value": 3,
    },
  ],
]
`;

exports[`o(Array) .splice() can insert and remove at the same time 3`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        3,
      ],
      "op": "splice",
      "target": Array [
        1,
        0,
        3,
      ],
      "value": Array [
        0,
      ],
    },
  ],
]
`;

exports[`o(Array) .splice() emits even if values did not change 1`] = `
Array [
  Array [
    Object {
      "key": 0,
      "oldValue": Array [
        1,
        2,
        3,
      ],
      "op": "splice",
      "target": Array [
        1,
        2,
        3,
      ],
      "value": Array [
        1,
        2,
        3,
      ],
    },
  ],
]
`;

exports[`o(Array) .splice() insertion can insert at the end 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [],
      "op": "splice",
      "target": Array [
        1,
        2,
        3,
      ],
      "value": Array [
        2,
        3,
      ],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 1,
      "op": "replace",
      "target": Array [
        1,
        2,
        3,
      ],
      "value": 3,
    },
  ],
]
`;

exports[`o(Array) .splice() insertion can insert at the start 1`] = `
Array [
  Array [
    Object {
      "key": 0,
      "oldValue": Array [],
      "op": "splice",
      "target": Array [
        3,
        2,
        1,
      ],
      "value": Array [
        3,
        2,
      ],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 1,
      "op": "replace",
      "target": Array [
        3,
        2,
        1,
      ],
      "value": 3,
    },
  ],
]
`;

exports[`o(Array) .splice() insertion can insert into the middle 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [],
      "op": "splice",
      "target": Array [
        1,
        0,
        0,
        2,
      ],
      "value": Array [
        0,
        0,
      ],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 2,
      "op": "replace",
      "target": Array [
        1,
        0,
        0,
        2,
      ],
      "value": 4,
    },
  ],
]
`;

exports[`o(Array) .splice() negative start index starts from the end 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        2,
      ],
      "op": "splice",
      "target": Array [
        1,
        3,
      ],
      "value": Array [],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 3,
      "op": "replace",
      "target": Array [
        1,
        3,
      ],
      "value": 2,
    },
  ],
]
`;

exports[`o(Array) .splice() negative start index stops at zero 1`] = `
Array [
  Array [
    Object {
      "key": 0,
      "oldValue": Array [
        1,
      ],
      "op": "splice",
      "target": Array [
        2,
      ],
      "value": Array [],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 2,
      "op": "replace",
      "target": Array [
        2,
      ],
      "value": 1,
    },
  ],
]
`;

exports[`o(Array) .splice() removal can splice the end 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        2,
        3,
      ],
      "op": "splice",
      "target": Array [
        1,
      ],
      "value": Array [],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 3,
      "op": "replace",
      "target": Array [
        1,
      ],
      "value": 1,
    },
  ],
]
`;

exports[`o(Array) .splice() removal can splice the middle 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        2,
        3,
      ],
      "op": "splice",
      "target": Array [
        1,
        4,
      ],
      "value": Array [],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 4,
      "op": "replace",
      "target": Array [
        1,
        4,
      ],
      "value": 2,
    },
  ],
]
`;

exports[`o(Array) .splice() removal can splice the start 1`] = `
Array [
  Array [
    Object {
      "key": 0,
      "oldValue": Array [
        1,
        2,
      ],
      "op": "splice",
      "target": Array [
        3,
      ],
      "value": Array [],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 3,
      "op": "replace",
      "target": Array [
        3,
      ],
      "value": 1,
    },
  ],
]
`;

exports[`o(Array) .splice() remove count can exceed the last index 1`] = `
Array [
  Array [
    Object {
      "key": 1,
      "oldValue": Array [
        2,
      ],
      "op": "splice",
      "target": Array [
        1,
      ],
      "value": Array [],
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 2,
      "op": "replace",
      "target": Array [
        1,
      ],
      "value": 1,
    },
  ],
]
`;

exports[`o(Array) [[Set]] emits for new indices 1`] = `
Array [
  Array [
    Object {
      "key": "1",
      "op": "add",
      "target": Array [
        undefined,
        1,
      ],
      "value": 1,
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 0,
      "op": "replace",
      "target": Array [
        undefined,
        1,
      ],
      "value": 2,
    },
  ],
]
`;

exports[`o(Array) [[Set]] emits for new values 1`] = `
Array [
  Array [
    Object {
      "key": "0",
      "oldValue": 0,
      "op": "replace",
      "target": Array [
        1,
      ],
      "value": 1,
    },
  ],
]
`;

exports[`o(Object) [[Delete]] emits for known keys 1`] = `
Array [
  Array [
    Object {
      "key": "a",
      "oldValue": 1,
      "op": "remove",
      "target": Object {},
    },
  ],
]
`;

exports[`o(Object) [[Set]] emits for new keys 1`] = `
Array [
  Array [
    Object {
      "key": "a",
      "op": "add",
      "target": Object {
        "a": 1,
      },
      "value": 1,
    },
  ],
]
`;

exports[`o(Object) [[Set]] emits for new values 1`] = `
Array [
  Array [
    Object {
      "key": "a",
      "oldValue": 0,
      "op": "replace",
      "target": Object {
        "a": 1,
      },
      "value": 1,
    },
  ],
]
`;

exports[`o(Set) .add() emits for new values 1`] = `
Array [
  Array [
    Object {
      "key": undefined,
      "op": "add",
      "target": Set {
        1,
      },
      "value": 1,
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 0,
      "op": "replace",
      "target": Set {
        1,
      },
      "value": 1,
    },
  ],
]
`;

exports[`o(Set) .clear() emits only when not empty 1`] = `
Array [
  Array [
    Object {
      "oldValue": Set {
        1,
        2,
      },
      "op": "clear",
      "target": Set {},
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 2,
      "op": "replace",
      "target": Set {},
      "value": 0,
    },
  ],
]
`;

exports[`o(Set) .delete() emits for known values 1`] = `
Array [
  Array [
    Object {
      "key": undefined,
      "oldValue": 1,
      "op": "remove",
      "target": Set {},
    },
  ],
  Array [
    Object {
      "key": Symbol(wana:size),
      "oldValue": 1,
      "op": "replace",
      "target": Set {},
      "value": 0,
    },
  ],
]
`;
